src/Compilers/Core/Portable/InternalUtilities/ExceptionUtilities.cs

```diff
@@ -27,5 +28,14 @@ internal static Exception Unreachable
         {
             get { return new InvalidOperationException("This program location is thought to be unreachable."); }
         }
+
+        /// <summary>
+        /// Determine if an exception was an <see cref="OperationCanceledException"/>, and that the provided token caused the cancellation.
+        /// </summary>
+        /// <param name="exception">The exception to test.</param>
+        /// <param name="cancellationToken">Checked to see if the provided token was cancelled.</param>
+        /// <returns><see langword="true"/> if the exception was an <see cref="OperationCanceledException" /> and the token was canceled.</returns>
+        internal static bool IsCurrentOperationBeingCancelled(Exception exception, CancellationToken cancellationToken)
+            => exception is OperationCanceledException oce && oce.CancellationToken == cancellationToken;
```

Comment:
❗ Do not check for cancellation token equivalence. The only thing that matters is that cancellationToken _also_ has cancellation requested.

Reply:
See also:

https://github.com/dotnet/roslyn/blob/0be180415eba3e45efc04bded886e29867956dcb/src/Compilers/Core/Portable/InternalUtilities/FatalError.cs#L94-L95

https://github.com/dotnet/roslyn/pull/41477#discussion_r376189922

Reply:
So I don't entirely get that, since the framework itself absolutely does cancellation token equality checks, for example when it's trying to decide to mark as Task as cancelled or faulted:

https://github.com/dotnet/runtime/blob/29cc2a2c23ce15469575a0b25c8b8b453a1d975f/src/libraries/System.Private.CoreLib/src/System/Threading/Tasks/Task.cs#L2409

I see the last time we had this conversation I asked for a follow up -- did we ever do that? :smile:

Reply:
For code using async/await, it ends up not mattering in practice.

Reply:
Here's code where async state machines are better aligned with Roslyn's current approach:
https://github.com/dotnet/runtime/blob/43ceed5514ca93242ff41f9f1ee53fc108c5dc1b/src/libraries/System.Private.CoreLib/src/System/Runtime/CompilerServices/AsyncTaskMethodBuilderT.cs#L471-L474